Better message for invalid types like List(int)

The error message "invalid type comment or annotation" isn't very helpful if a user writes `List(int)` instead of `List[int]` (or `Optional(int)`).  This seems to be a common error.

Example:

```py
from typing import List

def f():
    # type: () -> List(int)   # invalid type comment or annotation
    return [1]
```

A better message could be something like this:

```
program.py:3: error: Syntax error in type annotation
program.py:3:note: Suggestion: Use List[...] instead of List(...)
```

--- a/mypy/fastparse.py
+++ b/mypy/fastparse.py
@@ -1015,6 +1015,10 @@ class TypeConverter(ast3.NodeTransformer):
         if self.errors:
             self.errors.report(line, column, msg)
 
+    def note(self, msg: str, line: int, column: int) -> None:
+        if self.errors:
+            self.errors.report(line, column, msg, severity='note')
+
     def visit_raw_str(self, s: str) -> Type:
         # An escape hatch that allows the AST walker in fastparse2 to
         # directly hook into the Python 3.5 type converter in some cases
@@ -1031,12 +1035,19 @@ class TypeConverter(ast3.NodeTransformer):
 
     def visit_Call(self, e: ast3.Call) -> Type:
         # Parse the arg constructor
-        if not isinstance(self.parent(), ast3.List):
-            return self.generic_visit(e)
         f = e.func
         constructor = stringify_name(f)
+
+        if not isinstance(self.parent(), ast3.List):
+            self.fail(TYPE_COMMENT_AST_ERROR, self.line, e.col_offset)
+            if constructor:
+                self.note("Suggestion: use {}[...] instead of {}(...)".format(
+                    constructor, constructor),
+                    self.line, e.col_offset)
+            return AnyType(TypeOfAny.from_error)
         if not constructor:
             self.fail("Expected arg constructor name", e.lineno, e.col_offset)
+
         name = None  # type: Optional[str]
         default_type = AnyType(TypeOfAny.special_form)
         typ = default_type  # type: Type

--- a/test-data/unit/check-fastparse.test
+++ b/test-data/unit/check-fastparse.test
@@ -235,6 +235,30 @@ def f(a):
     pass
 [out]
 main:3: error: invalid type comment or annotation
+main:3: note: Suggestion: use Tuple[...] instead of Tuple(...)
+
+[case testFasterParseTypeErrorList_python2]
+
+from typing import List
+def f(a):
+    # type: (List(int)) -> int
+    pass
+[out]
+main:3: error: invalid type comment or annotation
+main:3: note: Suggestion: use List[...] instead of List(...)
+
+[case testFasterParseTypeErrorCustom]
+
+from typing import TypeVar, Generic
+T = TypeVar('T')
+class Foo(Generic[T]):
+  pass
+
+def f(a: Foo(int)) -> int:
+    pass
+[out]
+main:7: error: invalid type comment or annotation
+main:7: note: Suggestion: use Foo[...] instead of Foo(...)
 
 [case testFastParseMatMul]
 

